Group Ironic nodes to be associate with aggregates

gryf opened the pull request.

gryf committed:
Allow compute nodes to be associated with host agg patch

gryf commented:
@epim, @jimrollenhagen please, take a look.

jimrollenhagen commented on affinity_ocata/0001-allow-compute-nodes-to-be-associated-with-host-agg.patch:
```diff
@@ -0,0 +1,247 @@
++        try:
++            objects.Service.get_by_compute_host(ctx, host_or_node)
++            return True
++        except exception.HostMappingNotFound:
```
Is this the right error? I thought the host mapping was a cells thing.

gryf replied:
Yeah, you're right, should be `exception.ComputeHostNotFound` instead. I'm scratching my head though, since I have it right on my codebase… Will send it fixed soon

jimrollenhagen reviewed (commented):
Seems fine to me (assuming it applies cleanly and tests pass), just one small question inline.

gryf committed:
Fixed allow compute node patch

gryf committed:
Added another portion of Jay patches backported to Ocata

jimrollenhagen commented:
Looks pretty clean, though based on this comment, I assume tests don't pass? https://review.openstack.org/#/c/531243/1/nova/tests/functional/db/test_instance_group.py@286

Or does the way you suggest creating the instances solve the same problem?

gryf commented:
Passing project_id to instance object solved the issue.

gryf commented:
I'll change the comment, to be more clear about failing part. All nova tests (pep8, functional and py27) are passing now.

gryf committed:
Adds support for two new server group policies

This is an Ocata backport of Jay patch for adding support for two new
policies aggregate-affinity and aggregate-anti-affinity.

https://review.openstack.org/#/c/529201

epim reviewed (approved):
👍

jimrollenhagen reviewed (approved):
Move this stuff to a `nova/` directory instead of `affinity_ocata` and I'm good with merging it :)

gryf commented:
Please, wait a little bit - there are couple of fixes ongoing. I'll update patches soon.

gryf committed:
Added fixes and readme for the feature

gryf committed:
Added soft anti/affinity weight.

jimrollenhagen commented on affinity_ocata/README.rst:
```diff
@@ -0,0 +1,308 @@
+.. code:: ini
+
+   [filter_scheduler]
+   host_subset_size = 1
```
Had a conversation in Slack about this - @gryf has some ideas on fixing it and is going to write some code tomorrow.

```
18:51:17         jroll | hey y'all, affinity stuff is looking good so far
18:51:23         jroll | this concerns me a bit: https://github.com/yahoo/openstack/pull/2/files#diff-6d511686793305eaaaca7c10966bd46eR112 | Group Ironic nodes to be associate with aggregates by gryf · Pull Request #2 · yahoo/openstack
18:51:59         jroll | mostly due to this (from ironic docs): ```
18:51:59               | # Size of subset of best hosts selected by scheduler.
18:51:59               | # New instances will be scheduled on a host chosen randomly from a
18:51:59               | # subset of the 999 hosts. The big value is used to avoid race
18:51:59               | # conditions, when several instances are scheduled on the same bare
18:51:59               | # metal nodes. This is not a problem when resource classes are used.
18:51:59               | host_subset_size=999
18:51:59               | ```
18:53:00          gryf | yea. there is a link in commit message to the work, where someone need that
18:54:04          gryf | this thing is working like this - if host_subset_size is set to something larger than 1, than affinity (for aggregate or for vms) weight probably select wrong host
18:55:33          gryf | because logic behind this is to select not the best host which weight will provide, but random host from the set weight provide, even though all the other host might have negative weight.
18:55:59          gryf | devstack set this option to 999
18:56:08          gryf | I think that default is 1  (edited)
18:56:16         jroll | but also, because of an (undocumented, apparently?) problem with how ironic presents nodes to the scheduler - they're always in the same order, so in an environment where nodes are recycled often, the same node will get hit often for scheduling
18:56:25         jroll | leading to some nodes wearing out faster than others
18:56:31         jroll | (sorry, got pulled away for a minute)
18:56:51          gryf | yeah, that might be an issue
18:56:52         jroll | yeah, I understand the problem, I'm not sure the best route to fix it. but wanted to bring it up and get people thinking about it
18:57:28         jroll | this is only with soft affinity, right, because it's a weigher instead of a filter?
18:57:34          gryf | yes
18:57:43         jroll | ok cool
18:58:03          gryf | aggregate filters will remove hosts from the list
18:58:20         jroll | yep, makes sense, just double checking
18:59:00          gryf | I think, that this issue might be worked around by first selecting the best matches (with random pick) and if there is no suitable host, than select from the worst matches
19:00:07          gryf | I can cook up a quick fix for that tomorrow
19:03:29          gryf | I'm just wondering, that no one ever complain about how soft-affinity/soft-anti-affinity work after that patch, since even though there are compute hosts which can host the vms for affinity, the other compute will be selected…
19:03:41         jroll | @gryf: I'd be interested to see what you're thinking there, in code
19:04:03          gryf | ok than. I'll prepare some patch tomorrow
19:04:03         jroll | yeah, I agree the best behavior is with `host_subset_size=1`
19:04:21         jroll | feel free to just paste a diff in here if it's small, I just don't quite get what you plan to change
19:04:33          gryf | :+1:
19:04:45         jroll | thanks!
19:04:50          gryf | np :D
```

jimrollenhagen replied:
This is handled by an upstream commit in queens - https://review.openstack.org/#/c/494136/

Going to backport that internally but leave it off of this PR.

jimrollenhagen reviewed (approved):
Tested this; works as expected. Ready to merge this, but leaving it open to discuss the comment I left inline.

jimrollenhagen commented:
This is ready to merge but I don't have access :(

jimrollenhagen commented:
@gryf where'd the code go? I have it locally, but would be nice to merge this instead of dropping the code and closing it. I assume this was unintentional? :)

gryf commented:
oh shut. I'll revert the branch soon :/

gryf commented:
Done. Sorry for the inconvenience.

jimrollenhagen commented:
Not a problem, thanks!

sagarun commented:
@jimrollenhagen Do you have access to merge this? I can merge this if you want. Are you part of the Yahoo org?

jimrollenhagen commented:
@sagarun feel free to merge, I'm working on getting access but don't have it yet.
